kubespawner/utils.py

```diff
@@ -222,3 +223,45 @@ def recursive_update(target, new):
 
         else:
             target[k] = v
+
+
+class IgnoreMissing(dict):
+    """
+    Dictionary subclass for use with format_map
+
+    Renders missing values as "{key}", so format strings with
+    missing values just get rendered as is.
+
+    Stolen from https://docs.python.org/3/library/stdtypes.html#str.format_map
+    """
+
+    def __missing__(self, key):
+        return f"{{{key}}}"
+
+
+def recursive_format(
+    format_object: Union[str, dict, list], **kwargs
+) -> Union[str, dict, list]:
+    """
+    Recursively format given object with values provided as keyword arguments.
+
+    If the given object (string, list or dict) has items that do not have
+    placeholders for passed in kwargs, no formatting is performed.
+
+    recursive_format("{v}", v=5) -> Returns "5"
+    recrusive_format("{a}") -> Returns "{a}" rather than erroring, as is
+    the behavior of "format"
+    """
+    if isinstance(format_object, str):
+        return format_object.format_map(IgnoreMissing(kwargs))
+    elif isinstance(format_object, list):
+        return [recursive_format(i, **kwargs) for i in format_object]
+    elif isinstance(format_object, dict):
+        return {
+            recursive_format(k, **kwargs): recursive_format(v, **kwargs)
+            for k, v in format_object.items()
+        }
+    else:
+        raise ValueError(
```

Comment:
Is it reasonable for recursive_format to be called on a dict with non-string scalars, e.g. numbers? In that case, presumably they should be returned as-is rather than raise. I think this raise means we're still requiring all override values to be strings.

Reply:
@minrk ah good catch! The values can indeed be ints. I added a test, and relaxed this to not throw but instead return the object unformatted.